sim/mcuboot-sys/src/api.rs

```diff
@@ -7,97 +7,202 @@ use log::{Level, log_enabled, warn};
 use std::{
     collections::HashMap,
     mem,
-    ops::Deref,
     slice,
     sync::Mutex,
 };
+use crate::area::CAreaDesc;
+use std::thread;
+use std::ptr;
 
 /// A FlashMap maintain a table of [device_id -> Flash trait]
 pub type FlashMap = HashMap<u8, FlashPtr>;
 
+pub struct FlashParamsStruct {
+    align: u8,
+    erased_val: u8,
+}
+
+pub type FlashParams = HashMap<u8, FlashParamsStruct>;
+
+pub struct CAreaDescPtr {
+   pub ptr: *const CAreaDesc,
+}
+unsafe impl Send for CAreaDescPtr {}
+
+struct FlashContext {
+    flash_map: FlashMap,
+    flash_params: FlashParams,
+    flash_areas: CAreaDescPtr,
+}
+
+impl FlashContext {
+    pub fn new() -> FlashContext {
+        FlashContext {
+            flash_map: HashMap::new(),
+            flash_params: HashMap::new(),
+            flash_areas: CAreaDescPtr{ptr: ptr::null()},
+        }
+    }
+}
+
 lazy_static! {
-    static ref FLASH: Mutex<FlashMap> = {
+    static ref THREAD_CTX: Mutex<HashMap<thread::ThreadId, FlashContext>> = {
         Mutex::new(HashMap::new())
     };
 }
 
-struct FlashParams {
-    align: u8,
-    erased_val: u8,
+#[repr(C)]
+#[derive(Debug, Default)]
+pub struct CSimContext {
+    pub flash_counter: libc::c_int,
+    pub jumped: libc::c_int,
+    pub c_asserts: u8,
+    pub c_catch_asserts: u8,
+    // NOTE: Always leave boot_jmpbuf declaration at the end; this should
+    // store a "jmp_buf" which is arch specific and not defined by libc crate.
+    // The size below is enough to store data on a x86_64 machine.
+    pub boot_jmpbuf: [u64; 16],
+}
+
+pub struct CSimContextPtr {
+   pub ptr: *const CSimContext,
 }
+unsafe impl Send for CSimContextPtr {}
 
 lazy_static! {
-    static ref FLASH_PARAMS: Mutex<HashMap<u8, FlashParams>> = {
+    static ref SIM_CONTEXT: Mutex<HashMap<thread::ThreadId, CSimContextPtr>> = {
         Mutex::new(HashMap::new())
     };
 }
 
 // Set the flash device to be used by the simulation.  The pointer is unsafely stashed away.
 pub unsafe fn set_flash(dev_id: u8, dev: &mut dyn Flash) {
-    let mut flash_params = FLASH_PARAMS.lock().unwrap();
-    flash_params.insert(dev_id, FlashParams {
+    let mut map = THREAD_CTX.lock().unwrap();
+    let thread_id = thread::current().id();
+    if !map.contains_key(&thread_id) {
```

Comment:
(aside from this not being needed with thread local variables), using map.entry(...) should make this a bit cleaner.